File: backend/models/feed_post_model.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from datetime import datetime

from core.database import Base


class FeedPostModel(Base):
    """Campus Feed post — broadcasts from Student Council or admin."""

    __tablename__ = "feed_posts"

    id = Column(Integer, primary_key=True, index=True)
    tag = Column(String, nullable=False)                   # ANNOUNCEMENT, CAMPUS EVENT, CLUB NEWS, REMINDER
    title = Column(String, nullable=False, index=True)
    body = Column(String, nullable=False)
    image_url = Column(String, nullable=True)
    author_label = Column(String, default="Student Council")  # Display name
    seen_count = Column(Integer, default=0)
    is_pinned = Column(Boolean, default=False)
    author_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

File: frontend/lib/features/feed/data/models/feed_post_model.dart
class FeedPostModel {
  final int id;
  final String tag;
  final String title;
  final String body;
  final String? imageUrl;
  final String authorLabel;
  final int seenCount;
  final bool isPinned;
  final String timeAgo;
  final DateTime createdAt;

  FeedPostModel({
    required this.id,
    required this.tag,
    required this.title,
    required this.body,
    this.imageUrl,
    required this.authorLabel,
    required this.seenCount,
    required this.isPinned,
    required this.timeAgo,
    required this.createdAt,
  });

  factory FeedPostModel.fromJson(Map<String, dynamic> json) {
    return FeedPostModel(
      id: json['id'],
      tag: json['tag'],
      title: json['title'],
      body: json['body'],
      imageUrl: json['image_url'],
      authorLabel: json['author_label'] ?? 'Student Council',
      seenCount: json['seen_count'] ?? 0,
      isPinned: json['is_pinned'] ?? false,
      timeAgo: json['time_ago'] ?? '',
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'tag': tag,
      'title': title,
      'body': body,
      'image_url': imageUrl,
      'author_label': authorLabel,
      'seen_count': seenCount,
      'is_pinned': isPinned,
      'time_ago': timeAgo,
      'created_at': createdAt.toIso8601String(),
    };
  }
}
